### PR TITLE
DA7219 machine driver update

### PR DESCRIPTION
**Pull request by brentlu**
1. In preparation for ADL board support, I refactor the machine driver to use maxim-common module and remove redundant code. No functional change here and driver is tested on JSL board with MAX98360A amplifier.
2. Add adl_mx98360_da7219 board config to support DA7219 on ADL boards; tested on ADL-N chromebook

### CODE REVIEW

**Review by bardliao — DISMISSED** [review]

**Review by plbossart — DISMISSED** [review]

**Comment by brentlu** [comment]
Could we merge this PR? Thanks.

**Comment by plbossart** [comment]
Can you revisit the checkpatch warnings @brentlu?

````
WARNING: struct snd_soc_ops should normally be const
#159: FILE: sound/soc/intel/boards/sof_da7219.c:194:
+static struct snd_soc_ops max98373_ops = {

CHECK: braces {} should be used on all arms of this statement
#259: FILE: sound/soc/intel/boards/sof_da7219.c:368:
+	if (sof_da7219_quirk & SOF_MAX98360A_SPEAKER_AMP_PRESENT)
[...]
+	else if (sof_da7219_quirk & SOF_MAX98373_SPEAKER_AMP_PRESENT) {
[...]

````

**Comment by brentlu** [comment]
1. add const to max98373_ops
2. remove global variable 'headset'
3. add one commit for adl_mx98360_da7219 board; tested on ADL-N chromebook with topology https://github.com/thesofproject/sof/pull/8103

**Comment by brentlu** [comment]
1. fix the warning about brace

**Comment by brentlu** [comment]
rebase to see if it could pass the build test...

File: sound/soc/intel/boards/Kconfig [hunk 1]
```diff
@@ -606,25 +623,6 @@ config SND_SOC_INTEL_SOF_CML_RT1011_RT5682_MACH
 
 endif ## SND_SOC_SOF_COMETLAKE && SND_SOC_SOF_HDA_LINK
 
-if SND_SOC_SOF_JASPERLAKE
-
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig:614]: it would be simpler if you just removed the if JASPERLAKE and updated the select/depends below. You changed the order so it's a bit hard to review really.

File: sound/soc/intel/boards/sof_da7219.c [hunk 2]
```diff
@@ -15,11 +15,13 @@
 #include <sound/soc-acpi.h>
 #include "../../codecs/da7219.h"
 #include "hda_dsp_common.h"
+#include "sof_maxim_common.h"
+
+/* Speaker amp type */
+#define SOF_MAX98360A_SPEAKER_AMP_PRESENT	BIT(0)
+#define SOF_MAX98373_SPEAKER_AMP_PRESENT	BIT(1)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_da7219.c]: do you really need this bit if it can be detected with ACPI?

File: sound/soc/intel/boards/sof_da7219.c [hunk 3]
```diff
@@ -354,30 +495,57 @@ static int audio_probe(struct platform_device *pdev)
...
 
+	card_da7219.dai_link = dai_links;
+
 	INIT_LIST_HEAD(&ctx->hdmi_pcm_list);
 
+	/* Use PLL bypass mode if MCLK is available, be sure to set the
+	 * frequency of MCLK to 12.288 or 24.576MHz on topology side.
+	 */
+	if (board_quirk & SOF_DA7219_PLL_BYPASS)
+		ctx->pll_bypass = 1;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_da7219.c]: why is this needed? What not just test the MCLK and automatically bypass the PLL if it's 12.228 or a multiple?

**Review by plbossart — COMMENTED** [review]
I think this can be improved a bit @brentlu 

It's not all bad but you are doing things differently to the other recent PRs so the code is a bit inconsistent in the directions.

File: sound/soc/intel/boards/Kconfig (hunk 1 above)
Comment (brentlu) [reply sound/soc/intel/boards/Kconfig:614]: ok. I'll move it back.

File: sound/soc/intel/boards/sof_da7219.c (hunk 2 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c]: I'd like to do it later because I need to backport these patches back to older branch. Would like to minimize the porting.

File: sound/soc/intel/boards/sof_da7219.c (hunk 3 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c]: oh I can do it in codec init. let me upload a new version.

File: sound/soc/intel/boards/Kconfig [hunk 4]
```diff
@@ -606,25 +606,22 @@ config SND_SOC_INTEL_SOF_CML_RT1011_RT5682_MACH
...
-config SND_SOC_INTEL_SOF_DA7219_MAX98373_MACH
-	tristate "SOF with DA7219 and MAX98373/MAX98360A in I2S Mode"
+config SND_SOC_INTEL_SOF_DA7219_MACH
+	tristate "SOF with DA7219 codec in I2S Mode"
 	depends on I2C && ACPI
-	depends on MFD_INTEL_LPSS || COMPILE_TEST
-	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
-	select SND_SOC_INTEL_HDA_DSP_COMMON
+	depends on ((SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC) &&\
+		    (MFD_INTEL_LPSS || COMPILE_TEST))
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: i would keep the initial code

File: sound/soc/intel/boards/Kconfig [hunk 5]
```diff
@@ -606,25 +606,22 @@ config SND_SOC_INTEL_SOF_CML_RT1011_RT5682_MACH
...
 	select SND_SOC_DA7219
+	select SND_SOC_MAX98357A
 	select SND_SOC_MAX98373_I2C
 	select SND_SOC_DMIC
+	select SND_SOC_INTEL_HDA_DSP_COMMON
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: this is already in the initial code,

File: sound/soc/intel/boards/sof_da7219.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_da7219.c]: probably worth adding a comment then

**Review by plbossart — COMMENTED** [review]
only nit-picks from me...

**Review by bardliao — DISMISSED** [review]

File: sound/soc/intel/boards/sof_da7219.c [hunk 6]
```diff
@@ -395,10 +539,16 @@ static int audio_probe(struct platform_device *pdev)
...
+		.driver_data = (kernel_ulong_t)(SOF_MAX98373_SPEAKER_AMP_PRESENT |
+					SOF_DA7219_JSL_BOARD),
 	},
 	{
 		.name = "sof_da7219_mx98360a",
+		.driver_data = (kernel_ulong_t)(SOF_MAX98360A_SPEAKER_AMP_PRESENT |
+					SOF_DA7219_JSL_BOARD),
+	},
+	{
+		.name = "adl_mx98360_da7219",
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_da7219.c:561]: is there a reason why you dropped the 'sof' prefix?

File: sound/soc/intel/boards/sof_da7219.c [hunk 7]
```diff
@@ -539,17 +551,15 @@ static int audio_probe(struct platform_device *pdev)
...
 	{
 		.name = "sof_da7219_mx98360a",
-		.driver_data = (kernel_ulong_t)(SOF_MAX98360A_SPEAKER_AMP_PRESENT |
-					SOF_DA7219_JSL_BOARD),
+		.driver_data = (kernel_ulong_t)(SOF_DA7219_JSL_BOARD),
 	},
 	{
 		.name = "adl_mx98360_da7219",
-		.driver_data = (kernel_ulong_t)(SOF_MAX98360A_SPEAKER_AMP_PRESENT),
+		.driver_data = (kernel_ulong_t)(0),
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_da7219.c]: so do we need this quirk at all then?

**Review by plbossart — CHANGES_REQUESTED** [review]
lost on the last patch, sorry

File: sound/soc/intel/boards/sof_da7219.c (hunk 6 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c:561]: You mean the driver name "adl_mx98360_da7219"? I just follow the convention used in other machine drivers.

Two names "sof_da7219_mx98373" and "sof_da7219_mx98360a" were created for JSL boards long time ago so I didn't touch the names. Do you want me to change them, too?

File: sound/soc/intel/boards/sof_da7219.c (hunk 7 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c]: No quirk is needed for this board config now.

**Comment by brentlu** [comment]
The last patch just remove two quirks flags of amplifier type from board configs. No functional change there.

File: sound/soc/intel/boards/sof_da7219.c (hunk 7 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_da7219.c]: what happens if we just don't add .driver_data and add a comment to that effect?

File: sound/soc/intel/boards/sof_da7219.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_da7219.c:561]: that would be a nice clean-up, I am not sure what we have those two drivers that are named differently?

````
soc-acpi-intel-jsl-match.c\055:		.drv_name = "sof_da7219_mx98373",
soc-acpi-intel-jsl-match.c\062:		.drv_name = "sof_da7219_mx98360a",
````

File: sound/soc/intel/boards/Kconfig (hunk 4 above)
Comment (brentlu) [reply sound/soc/intel/boards/Kconfig]: fixed

File: sound/soc/intel/boards/Kconfig (hunk 5 above)
Comment (brentlu) [reply sound/soc/intel/boards/Kconfig]: restored

File: sound/soc/intel/boards/sof_da7219.c (hunk 6 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c:561]: driver names changed

File: sound/soc/intel/boards/sof_da7219.c (hunk 7 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_da7219.c]:
```
	unsigned long board_quirk = 0;
...
	if (pdev->id_entry && pdev->id_entry->driver_data)
		board_quirk = (unsigned long)pdev->id_entry->driver_data;
```

Since driver_data is zero (NULL), board_quirk will be the init value 0. It's working good on my ADL board.

**Review by plbossart — DISMISSED** [review]

**Review by bardliao — DISMISSED** [review]

**Comment by plbossart** [comment]
rebase needed @brentlu

**Comment by brentlu** [comment]
conflict in acpi enumeration table, resolved

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
@bardliao I'll let you merge this.

**Review by bardliao — APPROVED** [review]
